xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -63,19 +65,18 @@
    * be used to generate a service config.
    */
   static final class ConfigUpdate {
-    private final String clusterName;
     private final List<Route> routes;
 
-    private ConfigUpdate(String clusterName, List<Route> routes) {
-      this.clusterName = clusterName;
+    private ConfigUpdate(List<Route> routes) {
       this.routes = routes;
     }
 
+    /** Gets the cluster name for the default route. */
     String getClusterName() {
-      return clusterName;
+      return Iterables.getLast(routes).getRouteAction().getCluster();
```

Comment:
Ah... I was wondering how existing test cases in `XdsClientImplTest` are not affected, this is why. `ConfigUpdate` intends to be POJO, it's a data class. Its getter should not involve any business logic, it's just for getting a property value. This method doesn't make sense when `routes` contains more than one items and it should not be called in that case.

So I would still leaning towards `ConfigUpdate` containing a String and a List, with "oneof" semantics. XdsClient will set `clusterName` (String) if path matching is not enabled and leave `routes` (List) being `null`, and vice versa for path matching enabled.

Reply:
I agree that getter should not involve business logic, so I removed `getClusterName()`. It was just  used as a helper method, mostly for making test easy. I'm not a fan of oneof with two nullalbe fields either, because it will add more awkward business logic for `XdsNameResolver`.  Instead, I define a helper method in the test.

Reply:
The logic for not enabling path matching is wired, as you can see in XdsClientImpl, it goes into `Route` in `VirtualHost` to populate `clusterName` but it puts `Route` into `ConfigUpdate` and the resolver goes into the `Route` to populate `clusterName` again.

Reply:
> as you can see in XdsClientImpl, it goes into Route in VirtualHost to populate clusterName but it puts Route into ConfigUpdate and the resolver goes into the Route to populate clusterName again.

The XdsClientImple goes into Route in VirtualHost NOT to populate clusterName, it populates the whole route. (Yeah, it reads clusterName, but that's only for validation and INFO level logging.)

```java
    if (!enablePathMatching) {
        EnvoyProtoData.Route defaultRoute = Iterables.getLast(routes);
        configUpdate =
            ConfigUpdate.newBuilder()
                .addRoutes(ImmutableList.of(defaultRoute))
                .build();
        logger.log(
            XdsLogLevel.INFO,
            "Found cluster name (inlined in route config): {0}",
            defaultRoute.getRouteAction().getCluster());
      }
```

Reply:
The con of this approach is:
- For path matching not enabled case, `XdsNameResolver` needs to dig one level deeper to get `clusterName`.

The con of oneof approach is:
- The data structure of `ConfigUpdate` is ugly, and it is so only for the purpose of a usecase that might be migrated away.
- `XdsNameResolver` need to check null for the oneof fields.

Reply:
You have to populate `Route` message when path matching is disabled. It needs to validate

> whose match field must contain a prefix field whose value is the empty
string and whose route field must be set.

https://github.com/markdroth/proposal/commit/242375246285666fff622a60a988822c2466c09a#diff-05ea4a5894abbc0261b006741220598cR342-R344

This piece of logic is deleted in your last change, you really need to. revert it back.


Enabling/disabling patch matching runs in separate code paths, having them receiving different content of `ConfigUpdate` is a desired approach. IMO, using "oneof" semantics seems to be the desired way.

Reply:
> The data structure of `ConfigUpdate` is ugly, and it is so only for the purpose of a usecase that might be migrated away.

It's not "ugly". The two code paths use different fields in `ConfigUpdate`, separate code uses separate data, that's elegant. When eliminating one code path, just delete the field it uses.

> * `XdsNameResolver` need to check null for the oneof fields.

That's the beauty of null here. Elegant and concise meaning of something unset and do not use.

Reply:
> This piece of logic is deleted in your last change, you really need to. revert it back.

I still have the logic in `validateRoutes()`.

Reply:
I missed that one. But, still, I think using "oneof" different code paths makes implementation cleaner.

Reply:
I think we don't have to over-engineer for something that's temporary: the code path split for the enable path matching flag.
For me the code path split only in XdsClientImpl is easiest approach.

Reply:
Err... I don't think that's over-engineering, it's about implementation integrity and code style. Alright, I failed to convince you. I am not going to block you on this.